### PR TITLE
Live Chart Dynamo Implementation

### PR DESCRIPTION
**Pull request by dnenov**
### Purpose

This is a Dynamo Implementation of the [NodeModelCharts](https://github.com/alfarok/NodeModelCharts) from Keith Alfaro implements Graphing Utility Nodes.

The goal of this PR is to empower Dynamo users to better understand, work with, and visualize their data through interactive and responsive graphical charts that work both within the Dynamo workspace and allow propagation up through Dynamo Player.

#### before 
nodeLibrary

#### after 
final chart collection

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Release Notes

- Initial Release 
- Repository has been transferred to Dynamo
- Currently placed under the CoreNodeModelsWpf and CoreNode 
- CoreNodeModelWpf contains the Views and ViewModels
- CoreNode contains helper logic to allow output creation

### Reviewers

@reddyashish 
@mjkkirschner 

### FYIs

@Amoursol

### CODE REVIEW

**Comment by chuongmep** [comment]
Hi Team, Live Charts is old at the moment, can be consider use Live Charts 2 ?

File: src/Libraries/CoreNodeModels/CoreNodeModels.csproj [hunk 1]
```diff
@@ -10,6 +10,7 @@
     <AssemblyName>CoreNodeModels</AssemblyName>
     <OutputPath>$(OutputPath)\nodes\</OutputPath>
     <DocumentationFile>$(OutputPath)\CoreNodeModels.XML</DocumentationFile>
+    <UseWPF>true</UseWPF>
```
Comment (pinzart90) [inline src/Libraries/CoreNodeModels/CoreNodeModels.csproj:13]: I think this flag is supposed to be used in .net5 or above.
Also I do not think it is a good idea to bring in WPF/UI dependencies in a core project
That is why there is a CodeNodeModelsWpf project.

**Comment by Amoursol** [comment]
> Hi Team, Live Charts is old at the moment, can be consider use Live Charts 2 ?
> 
> https://github.com/beto-rodriguez/LiveCharts2

Hi Chuong! LiveCharts2 is indeed way more expansive, but due to the dependencies involved we are opting for the original for now as we aim to keep Dynamo's dependency landscape as small as possible.

File: src/Libraries/CoreNodeModels/CoreNodeModels.csproj (hunk 1 above)
Comment (dnenov) [reply src/Libraries/CoreNodeModels/CoreNodeModels.csproj:13]: Thank you for the comment. I will revert this change, but it was necessary for me to use some of the wpf functionality initially. Thank you!

**Review by sm6srw — CHANGES_REQUESTED** [review]
@dnenov I have done a partial review so far. Will continue tomorrow.

File: src/Libraries/CoreNodeModelsWpf/Charts/BarChartNodeModel.cs [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+        #region databridge
+        // Use the VMDataBridge to safely retrieve our input values
+
+        /// <summary>
+        /// Register the data bridge callback.
+        /// </summary>
+        protected override void OnBuilt()
+        {
+            base.OnBuilt();
+            VMDataBridge.DataBridge.Instance.RegisterCallback(GUID.ToString(), DataBridgeCallback);
```
[the remark this answers is not in the document]
Comment (dnenov) [reply src/Libraries/CoreNodeModelsWpf/Charts/BarChartNodeModel.cs:108]: Good call! It makes total sense, and it didn't blow up. :)

File: src/Libraries/CoreNodeModelsWpf/Charts/Controls/BarChartControl.xaml.cs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        {
+            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
+        }
+
+        public BarChartControl(BarChartNodeModel model)
+        {
+
+            InitializeComponent();
+
+            model.PropertyChanged += NodeModel_PropertyChanged;
```
[the remark this answers is not in the document]
Comment (dnenov) [reply src/Libraries/CoreNodeModelsWpf/Charts/Controls/BarChartControl.xaml.cs:32]: Absolutely.

File: src/Libraries/CoreNodeModelsWpf/Charts/BarChartNodeModel.cs [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+        {
+            InPorts.Add(new PortModel(PortType.Input, this, new PortData("labels", "A list of labels for the bar chart categories.")));
+            InPorts.Add(new PortModel(PortType.Input, this, new PortData("values", "A list (of lists) to supply values for the bars in each category.")));
+            InPorts.Add(new PortModel(PortType.Input, this, new PortData("colors", "A list of colors for each bar chart category.")));
+
+            OutPorts.Add(new PortModel(PortType.Output, this, new PortData("labels:values", "Dictionary containing label:value key-pairs")));
+
+            RegisterAllPorts();
+
+            PortDisconnected += PieChartNodeModel_PortDisconnected;
```
[the remark this answers is not in the document]
Comment (dnenov) [reply src/Libraries/CoreNodeModelsWpf/Charts/BarChartNodeModel.cs:69]: Yes, absolutely! There was a moment where I thought of addressing that, but then got sidetracked, thank you for spotting that out!

File: src/Libraries/CoreNodeModelsWpf/Charts/Controls/BarChartControl.xaml.cs [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+                }
+            }
+        }
+
+        /// <summary>
+        /// Unsubscribes from ViewModel events
+        /// </summary>
+        public void Dispose()
+        {
+            this.model.PropertyChanged -= NodeModel_PropertyChanged;
```
Comment (sm6srw) [inline src/Libraries/CoreNodeModelsWpf/Charts/Controls/BarChartControl.xaml.cs:151]: Is this Dispose function actually called? 

If not, I think you need to do something like this for user controls:

In the contructor:
`Unloaded += Unload;`

And then:
```
private void Unload(object sender, EventArgs e)
{
    this.model.PropertyChanged -= NodeModel_PropertyChanged;
    Unloaded -= Unload;
 }
```

File: src/Libraries/CoreNodeModelsWpf/CoreNodeModelsWpf - Backup.csproj [hunk 6]
```diff
@@ -0,0 +1,155 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Comment (sm6srw) [inline src/Libraries/CoreNodeModelsWpf/CoreNodeModelsWpf - Backup.csproj:1]: I guess that this file should be removed?

File: src/Libraries/CoreNodes/ChartHelpers/BarChartFunctions.cs [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+namespace CoreNodes.ChartHelpers
+{
+    public class BarChartFunctions
+    {
+        private BarChartFunctions() { }
+
+        [IsVisibleInDynamoLibrary(false)]
+        public static Dictionary<string, List<double>> GetNodeInput(List<string> labels, List<List<double>> values, List<DSCore.Color> colors)
+        {
+            // TODO - just pass input data unmodified instead?
```
Comment (sm6srw) [inline src/Libraries/CoreNodes/ChartHelpers/BarChartFunctions.cs:17]: Do you want to do something about this TODO? If not, remove it.

File: src/Libraries/CoreNodes/ChartHelpers/BarChartFunctions.cs [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+                output.Add(labels[i], values[i]);
+            }
+
+            return output;
+        }
+
+        [IsVisibleInDynamoLibrary(false)]
+        public static Dictionary<string, double> GetNodeInput(List<string> labels, List<double> values, List<DSCore.Color> colors)
+        {
+            // TODO - just pass input data unmodified instead?
```
Comment (sm6srw) [inline src/Libraries/CoreNodes/ChartHelpers/BarChartFunctions.cs:36]: Same here.

File: src/Libraries/CoreNodes/ChartHelpers/BasicLineChartFunctions.cs [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+namespace CoreNodes.ChartHelpers
+{
+    public class BasicLineChartFunctions
+    {
+        private BasicLineChartFunctions() { }
+
+        [IsVisibleInDynamoLibrary(false)]
+        public static Dictionary<string, List<double>> GetNodeInput(List<string> titles, List<List<double>> values, List<DSCore.Color> colors)
+        {
+            // TODO - just pass input data unmodified instead?
```
Comment (sm6srw) [inline src/Libraries/CoreNodes/ChartHelpers/BasicLineChartFunctions.cs:17]: TODO?

File: src/Libraries/CoreNodes/ChartHelpers/HeatSeriesFunctions.cs [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+namespace CoreNodes.ChartHelpers
+{
+    public class HeatSeriesFunctions
+    {
+        private HeatSeriesFunctions() { }
+
+        [IsVisibleInDynamoLibrary(false)]
+        public static Dictionary<string, List<double>> GetNodeInput(List<string> xLabels, List<string> yLabels, List<List<double>> values, List<DSCore.Color> colors)
+        {
+            // TODO - just pass input data unmodified instead?
```
Comment (sm6srw) [inline src/Libraries/CoreNodes/ChartHelpers/HeatSeriesFunctions.cs:17]: TODO?

File: src/Libraries/CoreNodes/ChartHelpers/PieChartFunctions.cs [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        private PieChartFunctions()
+        {
+        }
+
+        [IsVisibleInDynamoLibrary(false)]
+        public static Dictionary<string, double> GetNodeInput(List<string> labels, List<double> values,
+            List<DSCore.Color> colors)
+        {
+            // TODO - just pass input data unmodified instead?
```
Comment (sm6srw) [inline src/Libraries/CoreNodes/ChartHelpers/PieChartFunctions.cs:20]: TODO?

File: src/Libraries/CoreNodes/ChartHelpers/ScatterPlotFunctions.cs [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+namespace CoreNodes.ChartHelpers
+{
+    public class ScatterPlotFunctions
+    {
+        private ScatterPlotFunctions() { }
+
+        [IsVisibleInDynamoLibrary(false)]
+        public static Dictionary<string, Dictionary<string, List<double>>> GetNodeInput(List<string> titles, List<List<double>> xValues, List<List<double>> yValues, List<DSCore.Color> colors)
+        {
+            // TODO - just pass input data unmodified instead?
```
Comment (sm6srw) [inline src/Libraries/CoreNodes/ChartHelpers/ScatterPlotFunctions.cs:17]: TODO?

File: src/Libraries/CoreNodes/ChartHelpers/XYLineChartFunctions.cs [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+namespace CoreNodes.ChartHelpers
+{
+    public class XYLineChartFunctions
+    {
+        private XYLineChartFunctions() { }
+
+        [IsVisibleInDynamoLibrary(false)]
+        public static Dictionary<string, Dictionary<string, List<double>>> GetNodeInput(List<string> titles, List<List<double>> xValues, List<List<double>> yValues, List<DSCore.Color> colors)
+        {
+            // TODO - just pass input data unmodified instead?
```
Comment (sm6srw) [inline src/Libraries/CoreNodes/ChartHelpers/XYLineChartFunctions.cs:17]: TODO?

**Review by sm6srw — CHANGES_REQUESTED** [review]
We are getting close. I have a question about the memory management in user controls and there is a couple of TODO's that needs to be addressed.

File: src/Libraries/CoreNodeModelsWpf/CoreNodeModelsWpf - Backup.csproj (hunk 6 above)
Comment (dnenov) [reply src/Libraries/CoreNodeModelsWpf/CoreNodeModelsWpf - Backup.csproj:1]: Yup, absolutely. Removed it.

**Review by sm6srw — COMMENTED** [review]
Looks good. I will kickoff the tests again just to make sure.

**Review by sm6srw — APPROVED** [review]
Tests passed here: https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/8285/

This is now ready to me merged.

**Pull request merged**